gnuradio-runtime/python/gnuradio/gr/gateway.py

```diff
@@ -77,36 +77,62 @@ def eval(self, arg):
             raise ex
         return 0
 
+########################################################################
+# io_signature for Python
+########################################################################
+class py_io_signature(object):
+
+    # Minimum and maximum number of ports, and a list of numpy types.
+    def __init__(self, min_ports, max_ports, type_list):
+        self.__min_ports = min_ports
+        self.__max_ports = max_ports
+        self.__types = map(numpy.dtype, type_list)
+
+    # Make/return a gr.io_signature. A non-empty list of sizes is
+    # required, even if there are no ports.
+    def gr_io_signature(self):
+        return io_signaturev(self.__min_ports, self.__max_ports,
+                             [t.itemsize for t in self.__types] or [0])
+
+    # Return data types for the first nports ports. If nports is
+    # smaller than the provided type list, return a truncated list. If
+    # larger, fill with the last type.
+    def port_types(self, nports):
+        ntypes = len(self.__types)
+        if ntypes == 0:
+            return []
+        if nports <= ntypes:
+            return self.__types[:nports]
+        return self.__types + [self.__types[-1]]*(nports-ntypes)
+
 ########################################################################
 # The guts that make this into a gr block
 ########################################################################
 class gateway_block(object):
 
     def __init__(self, name, in_sig, out_sig, work_type, factor):
 
-        #ensure that the sigs are iterable dtypes
-        def sig_to_dtype_sig(sig):
-            if sig is None: sig = ()
-            return map(numpy.dtype, sig)
-        self.__in_sig = sig_to_dtype_sig(in_sig)
-        self.__out_sig = sig_to_dtype_sig(out_sig)
+        # Normalize the many Python ways of saying 'nothing' to '[]'
+        in_sig = in_sig or []
+        out_sig = out_sig or []
```

Comment:
Do we need a mutable type here? Would a tuple `()` be sufficient?

Reply:
mutable is not needed